test/common/router/router_test.cc

```diff
@@ -4454,17 +4458,118 @@ TEST_F(RouterTestChildSpan, ResetFlow) {
   EXPECT_CALL(callbacks_.active_span_, spawnChild_(_, "router fake_cluster egress", _))
       .WillOnce(Return(child_span));
   EXPECT_CALL(callbacks_, tracingConfig());
-  EXPECT_CALL(*child_span,
-              setTag(Eq(Tracing::Tags::get().Component), Eq(Tracing::Tags::get().Proxy)));
   router_.decodeHeaders(headers, true);
 
   Http::HeaderMapPtr response_headers(new Http::TestHeaderMapImpl{{":status", "200"}});
   response_decoder->decodeHeaders(std::move(response_headers), false);
 
+  EXPECT_CALL(*child_span,
+              setTag(Eq(Tracing::Tags::get().Component), Eq(Tracing::Tags::get().Proxy)));
+  EXPECT_CALL(*child_span, setTag(Eq(Tracing::Tags::get().HttpProtocol), Eq("HTTP/1.0")));
+  EXPECT_CALL(*child_span, setTag(Eq(Tracing::Tags::get().UpstreamCluster), Eq("fake_cluster")));
+  EXPECT_CALL(*child_span, setTag(Eq(Tracing::Tags::get().HttpStatusCode), Eq("200")));
+  EXPECT_CALL(*child_span, setTag(Eq(Tracing::Tags::get().ResponseFlags), Eq("UR")));
+  EXPECT_CALL(*child_span, setTag(Eq(Tracing::Tags::get().Error), Eq(Tracing::Tags::get().True)));
+  EXPECT_CALL(*child_span, setTag(Eq(Tracing::Tags::get().ErrorReason), Eq("remote reset")));
   EXPECT_CALL(*child_span, finishSpan());
   encoder.stream_.resetStream(Http::StreamResetReason::RemoteReset);
 }
 
+// Make sure child spans start/inject/finish with retry and cancellation flow.
+// First request times out and is retried, and then a response is received.
+// Second request is then cancelled: we don't attempt to retry because we already retried for
+// timeout.
+TEST_F(RouterTestChildSpan, RetryOnlyOnceForSameUpstreamRequest) {
```

Comment:
The span tags look consistent with the description - although my preference would be to simplify the scenario into separate tests for cancel and retry - although will defer to others whether this combined test is better.

Reply:
+1 agree, please split into 2 tests. The tests in this file are all super complicated to understand so if we can simplify that would be great.

/wait

Reply:
I added another test specifically for retries (no cancellation). But I can't figure out how to trigger a cancellation without having to retry a request. None of the other tests trigger a cancellation on the upstream request. If I can't do this, I'm not sure how to split this into a cancellation-only test. Please advise.

Reply:
A downstream reset should trigger an upstream reset/cancellation.

Reply:
Done. I simulated the cancellation using `router_.onDestroy()` because I couldn't figure out how to reset the upstream connection.